# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [3]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np

In [4]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0         Bulbasaur    Grass  Poison  45      49       49       65       65   
1           Ivysaur    Grass  Poison  60      62       63       80       80   
2          Venusaur    Grass  Poison  80      82       83      100      100   
3     Mega Venusaur    Grass  Poison  80     100      123      122      120   
4        Charmander     Fire     NaN  39      52       43       60       50   
..              ...      ...     ...  ..     ...      ...      ...      ...   
795         Diancie     Rock   Fairy  50     100      150      100      150   
796    Mega Diancie     Rock   Fairy  50     160      110      160      110   
797  Hoopa Confined  Psychic   Ghost  80     110       60      150      130   
798   Hoopa Unbound  Psychic    Dark  80     160       60      170      130   
799       Volcanion     Fire   Water  80     110      120      130       90   

     Speed  Generation  Legendary  
0       45           1      False  
1       60           1      False  
2       80           1      False  
3       80           1      False  
4       65           1      False  
..     ...         ...        ...  
795     50           6       True  
796    110           6       True  
797     70           6       True  
798     80           6       True  
799     70           6       True  

[800 rows x 11 columns]

- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

In [5]:
df_dragon = df[df["Type 1"]=="Dragon"]["HP"]
df_grass = df[df["Type 1"]=="Grass"]["HP"]

In [6]:
#Set the hypothesis

#H0: mu_hp dragon > mu_hp grass
#H1: mu_hp dragon < mu_hp grass

#significance level = 0.05

st.ttest_ind(df_dragon,df_grass, alternative='less')

TtestResult(statistic=3.590444254130357, pvalue=0.9997432030849847, df=100.0)

**Findings** : since the pvalue is higher than the significant level, we accept the hypothesis that Pokemons of type Dragon have on average more HP stats than Grass.

- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


In [14]:
import pandas as pd
from scipy.stats import f_oneway

#Set the hypothesis

#H0: legendary_pokemon_stats = non_legendary_pokemon_stats
#H0: legendary_pokemon_stats != non_legendary_pokemon_stats

#Lets choose significance level of 5%
alpha = 0.05

# List of stats columns
stats_columns = ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']

# Perform ANOVA test for each stat
for stat in stats_columns:
    legendary_pokemon = df[df['Legendary'] == True][stat]
    non_legendary_pokemon = df[df['Legendary'] == False][stat]
    
    # Perform one-way ANOVA test
    f_statistic, p_value = f_oneway(legendary_pokemon, non_legendary_pokemon)
    
    # Print results
    print(f'Stat: {stat}')
    print(f'F-statistic: {f_statistic:.2f}')
    print(f'p-value: {p_value:.4f}')
    if p_value < 0.05:
        print('Result: Reject null hypothesis - Significant difference found')
    else:
        print('Result: Fail to reject null hypothesis - No significant difference found')
    print('-' * 40)


Stat: HP
F-statistic: 64.58
p-value: 0.0000
Result: Reject null hypothesis - Significant difference found
----------------------------------------
Stat: Attack
F-statistic: 108.10
p-value: 0.0000
Result: Reject null hypothesis - Significant difference found
----------------------------------------
Stat: Defense
F-statistic: 51.57
p-value: 0.0000
Result: Reject null hypothesis - Significant difference found
----------------------------------------
Stat: Sp. Atk
F-statistic: 201.40
p-value: 0.0000
Result: Reject null hypothesis - Significant difference found
----------------------------------------
Stat: Sp. Def
F-statistic: 121.83
p-value: 0.0000
Result: Reject null hypothesis - Significant difference found
----------------------------------------
Stat: Speed
F-statistic: 95.36
p-value: 0.0000
Result: Reject null hypothesis - Significant difference found
----------------------------------------


Based on the p-values obtained, which is less than 0.05, we reject the null hypothesis and conclude that there is a statistically significant difference in the mean stats between Legendary and Non-Legendary Pokémon.

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [15]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0

**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 37)
- Hospital coordinates (-122, 34)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

In [55]:
# Let's calculate the distance for each school and each hospital : 

# distances to school and hospital
df['distance_to_school'] = ((df['longitude'] - school_coords[0]) ** 2 + (df['latitude'] - school_coords[1]) ** 2) ** 0.5
df['distance_to_hospital'] = ((df['longitude'] - hospital_coords[0]) ** 2 + (df['latitude'] - hospital_coords[1]) ** 2) ** 0.5


In [51]:
# Divide dataset into houses close and far from either a hospital or school

close_houses = df[(df['distance_to_school'] < threshold_distance) | (df['distance_to_hospital'] < threshold_distance)]
far_houses = df[(df['distance_to_school'] >= threshold_distance) & (df['distance_to_hospital'] >= threshold_distance)]

In [52]:
# Display the dataframe with houses close from either a hospital or school : 
display(close_houses)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
4523    -118.05     36.64                34.0       2090.0           478.0   
5596    -118.18     37.35                16.0       3806.0           794.0   
5597    -118.18     36.63                23.0       2311.0           487.0   
6776    -118.30     37.17                22.0       3480.0           673.0   
6904    -118.31     36.94                35.0       2563.0           530.0   

      population  households  median_income  median_house_value  \
4523       896.0       426.0         2.0357             74200.0   
5596      1501.0       714.0         2.1212            108300.0   
5597      1019.0       384.0         2.2574            104700.0   
6776      1541.0       636.0         2.7500             94500.0   
6904       861.0       371.0         2.3250             80600.0   

      distance_to_school  distance_to_hospital  
4523            0.363456              4.751010  
5596            0.393573              5.080837  
5597            0.411461              4.637812  
6776            0.344819              4.872258  
6904            0.315753              4.718019

In [53]:
# Display the dataframe with houses far from either a hospital or school : 
display(far_houses)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
2        -114.56     33.69                17.0        720.0           174.0   
3        -114.57     33.64                14.0       1501.0           337.0   
4        -114.57     33.57                20.0       1454.0           326.0   
...          ...       ...                 ...          ...             ...   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  \
0          1015.0       472.0         1.4936             66900.0   
1          1129.0       463.0         1.8200             80100.0   
2           333.0       117.0         1.6509             85700.0   
3           515.0       226.0         3.1917             73400.0   
4           624.0       262.0         1.9250             65500.0   
...           ...         ...            ...                 ...   
16995       907.0       369.0         2.3571            111400.0   
16996      1194.0       465.0         2.5179             79000.0   
16997      1244.0       456.0         3.0313            103600.0   
16998      1298.0       478.0         1.9797             85800.0   
16999       806.0       270.0         3.0147             94600.0   

       distance_to_school  distance_to_hospital  
0                4.638125              7.692347  
1                4.384165              7.540617  
2                4.773856              7.446456  
3                4.801510              7.438716  
4                4.850753              7.442432  
...                   ...                   ...  
16995            7.211380              6.957298  
16996            7.275232              7.064630  
16997            7.944533              8.170410  
16998            7.920227              8.132035  
16999            7.270083              6.949396  

[16995 rows x 11 columns]

In [54]:
# Choose the propper test and, with 5% significance, comment your findings.

from scipy.stats import mannwhitneyu

# Let's perform Mann-Whitney U test
statistic, p_value = mannwhitneyu(close_houses['median_house_value'], far_houses['median_house_value'])

alpha = 0.05

print("Mann-Whitney U test p-value:", p_value)

if p_value < alpha:
    print("Reject the null hypothesis. There is sufficient evidence to suggest that houses close to either a school or a hospital are more expensive.")
else:
    print("Fail to reject the null hypothesis. There is not enough evidence to suggest that houses close to either a school or a hospital are more expensive.")


Mann-Whitney U test p-value: 0.005835992784005986
Reject the null hypothesis. There is sufficient evidence to suggest that houses close to either a school or a hospital are more expensive.
